File: Nian_Karl/QL/src/ast/Value.java
package ast;

import ast.value.Bool;
import ast.value.Int;
import ast.value.Undefined;

public abstract class Value {
	
	private static final Undefined UNDEFINED = new Undefined(); 
	
	public abstract Object getValue();
	public abstract String toString();

	// Unary Expression
	public Value not() {return UNDEFINED;}
	public Value positive() {return UNDEFINED;}
	public Value negative() {return UNDEFINED;}
	
	// Binary Expression
	public Value plus(Value arg) {return UNDEFINED;}
	public Value minus(Value arg) {return UNDEFINED;}
	public Value multiply(Value arg) {return UNDEFINED;}
	public Value divide(Value arg) {return UNDEFINED;}
	public Value and(Value arg) {return UNDEFINED;}
	public Value or(Value arg) {return UNDEFINED;}
	public Value equal(Value arg) {return UNDEFINED;}
	public Value notEqual(Value arg) {return UNDEFINED;}
	public Value greater(Value arg) {return UNDEFINED;}
	public Value greaterEqual(Value arg) {return UNDEFINED;}
	public Value less(Value arg) {return UNDEFINED;}
	public Value lessEqual(Value arg) {return UNDEFINED;}
	
	// Double Dispatch
	public Value intPlus(Int arg) {return UNDEFINED;}
	public Value intMinus(Int arg) {return UNDEFINED;}
	public Value intMultiply(Int arg) {return UNDEFINED;}
	public Value intDivide(Int arg) {return UNDEFINED;}
	public Value boolAnd(Bool arg) {return UNDEFINED;}
	public Value boolOr(Bool arg) {return UNDEFINED;}
	public Value boolEqual(Bool arg) {return UNDEFINED;}
	public Value boolNotEqual(Bool arg) {return UNDEFINED;}
	public Value intGreater(Int arg) {return UNDEFINED;}
	public Value intGreaterEqual(Int arg) {return UNDEFINED;}
	public Value intLess(Int arg) {return UNDEFINED;}
	public Value intLessEqual(Int arg) {return UNDEFINED;}
	
}

File: FelipezConde/QuestionnaireLanguage/AST/Nodes/Values/Value.cs
using AST.Nodes.Interfaces;
using AST.Representation;
using AST.Storage;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Types = AST.Types;

namespace AST.Nodes.Values
{
    public abstract class Value
    {
        private readonly PositionInText Position;
        protected Value() { }
        protected Value(PositionInText position)
        {
            Position = position;
        }
        public virtual PositionInText GetPosition(){ return Position; }

        public abstract object GetType(ISymbolTable lookup);
        public abstract Value Equal(Value value);
        public abstract Value NotEqual(Value value);

        public virtual Bool Negate() { return null; }
        public virtual Value And(Value value) { return null; }
        public virtual Value Or(Value value) { return null; }
        public virtual Value Greater(Value value) { return null; }
        public virtual Value GreaterEqual(Value value) { return null; }
        public virtual Value Less(Value value) { return null; }
        public virtual Value LessEqual(Value value) { return null; }
        public virtual Value Add(Value value) { return null; }
        public virtual Value Substract(Value value) { return null; }
        public virtual Value Divide(Value value) { return null; }
        public virtual Value Multiply(Value value) { return null; }


        public virtual Value BoolAnd(Bool boolValue) { return null; }
        public virtual Value BoolOr(Bool boolValue) { return null; }
        public virtual Value BoolEqual(Bool boolValue) { return null; }
        public virtual Value BoolNotEqual(Bool boolValue) { return null; }

        public virtual Value StringEqual(Values.String stringValue) { return null; }
        public virtual Value StringNotEqual(Values.String stringValue) { return null; }

        public virtual Value IntegerAdd(Values.Int intValue) { return null; }
        public virtual Value IntegerSubstract(Values.Int intValue) { return null; }
        public virtual Value IntegerDivide(Values.Int intValue) { return null; }
        public virtual Value IntegerMultiply(Values.Int intValue) { return null; }
        public virtual Value IntegerEqual(Values.Int intValue) { return null; }
        public virtual Value IntegerNotEqual(Values.Int intValue) { return null; }
        public virtual Value IntegerGreater(Values.Int intValue) { return null; }
        public virtual Value IntegerGreaterEqual(Values.Int intValue) { return null; }
        public virtual Value IntegerLess(Values.Int intValue) { return null; }
        public virtual Value IntegerLessEqual(Values.Int intValue) { return null; }

        public virtual void Accept(Visitors.IVisitor visitor) {}

        public virtual T Accept<T>(Visitors.IVisitor<T> visitor) { return default(T); }

        public abstract string MakeString();

        public bool IsOfType(Types.Type type)
        {
            throw new NotImplementedException();
        }
    } 
}
